Blog Post: Strimzi Metrics Reporter (#531)
strimzi/strimzi.github.io: _posts/2025-08-26-strimzi-metrics-reporter.md

---
layout: post
title: "Using Strimzi Metrics Reporter to Expose Kafka Metrics in Prometheus Format"
date: 2025-10-01
author: Owen Corrigan
---
When deploying new Kafka cluster it may be one of your intentions to monitor the cluster using metrics.
`Apache Kafka®` generates and exposes metrics that reflect how a system is operating. These metrics are 
useful for monitoring, troubleshooting, tuning, and capacity planning when it comes to running your Kafka cluster.
In essence, monitoring is crucial to ensure the health and performance of your Kafka clusters.
`Prometheus` as a metrics monitoring solution has become increasingly common due to its popularity in the cloud-native ecosystem and it is this format that we use to expose metrics.

> The [Strimzi Metrics Reporter](https://github.com/strimzi/metrics-reporter) project implements the Java `MetricsReporter` interface to expose metrics via `Prometheus` and it is available from Strimzi v0.48.0.

Reviewer: ```suggestion
The [Strimzi Metrics Reporter](https://github.com/strimzi/metrics-reporter) project implements the Java `MetricsReporter` interface to expose metrics via `Prometheus` and it is available since 0.48.0 release.
```

Up to this point, we have recommended using the `JMX Exporter` as a way to expose Kafka metrics, and we have provided some example yaml files in our [Strimzi Examples folder](https://github.com/strimzi/strimzi-kafka-operator/tree/main/packaging/examples/metrics).

> However, using the JMX Exporter adds unnecessary complexity. Metrics reporters first expose metrics through JMX, and then a Java agent re-exposes them over HTTP.

Reviewer: Here you should mention the need to run the Prometheus JMX Eporter Java agent with its complex mapping rules, and load the Kafka JmxReporter plugin. This creates a significant overhead.

Author: Is it better now?

> The `Strimzi Metrics Reporter` exposes metrics in the Prometheus format directly over HTTP, simplifying the setup and reducing resource consumption.

Reviewer: Should we also mention that this is a `MetricsReporter` plugin that can be used with any Kafka deployment?

Here, we will discuss Strimzi Metrics Reporter installation and configuration, and 
guide you on how to use it effectively to monitor your Kafka clusters.

## Key Features
* Direct Prometheus Integration: Exposes Kafka metrics directly to Prometheus through an HTTP endpoint.
* Configurable Metrics Collection: Allows users to specify which metrics should be collected using a flexible allowlist.

### Deploying Metrics Reporter
The first step is to install the metrics reporter in your `Kafka Custom Resource`. To do this, add the following:

```yaml
apiVersion: kafka.strimzi.io/v1beta1
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
  # ...
    metricsConfig:
      type: strimziMetricsReporter
  # ...  
```

> By adding `type: strimziMetricsReporter` to the `metricsConfig` section of your Kafka CR, `Strimzi` will deploy a sensible set of default metric values.

Reviewer: ```suggestion
By adding `type: strimziMetricsReporter` to the `metricsConfig` section of your Kafka CR, `Strimzi` will export a sensible set of default metrics.
```

Author: Refactored

> However, you can add you own custom values (providing the metrics are available) by filtering the metrics by name. This is achieved by adding another field called `values` and within that field, adding `allowList`. Then you can specify which metrics you want to collect as a list. For example:

Reviewer: This is not a metrics list, but rather a regex list, where each entry is used to filter allowed metrics.

Author: Refactored

```yaml
apiVersion: kafka.strimzi.io/v1beta1
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
  # ...
    metricsConfig:
      type: strimziMetricsReporter
      values:
        allowList:
          - "kafka_log.*"
          - "kafka_network.*"
  # ...  
```

If you are adding Metrics Reporter to an existing cluster (instead of starting a fresh cluster) then you will see all your
Kafka broker pods roll so that the metrics reporter config can be added to your Kafka CR.

If you would like to deploy your Kafka cluster with the Metrics Reporter enabled from the start, you can use the examples in our [Strimzi Metrics Reporter Examples folder](https://github.com/strimzi/strimzi-kafka-operator/tree/main/packaging/examples/metrics/strimzi-metrics-reporter).
by running the following command from directory:

```bash
$ kubectl apply -f kafka-metrics.yaml -n myproject
```

We have also provided some [Grafana Dashboards](https://github.com/strimzi/strimzi-kafka-operator/tree/main/packaging/examples/metrics/strimzi-metrics-reporter/grafana-dashboards) in the same directory to help you get started with visualizing your Kafka metrics in the Prometheus format.

## Conclusion

> We have shown you how to deploy the `Strimzi Metrics Reporter` in your Kafka cluster and given some reasons why you might want to use it in preference to the previously recommended `JMX Exporter`.

Reviewer: I think we can do better to convince people to give it a chance. Feel free to use the simple benchmark from our StrimziCon presentation. Also, there is a recent user feedback that maybe we can use without mentioning the author for privacy reasons:

> That strimzi metrics Reporter just shaved about 1.5 G of memory usage on our kafka pods. Good job :heart:

The [Strimzi Metrics Reporter](https://github.com/strimzi/metrics-reporter) is a new project in Strimzi and it is available since Strimzi 0.48.0 release.
It implements the Kafka MetricsReporter interface to expose metrics via Prometheus.
Strimzi Metrics Reporter is designed from scratch as a Kafka plugin and allows you to configure it directly from the Kafka configuration.
That also means it is more efficient compared to other monitoring tools that run as sidecars or JVM agents.
Strimzi Metrics Reporter has fixed names for all metrics giving users much less flexibility than other tools which provide complex mapping rules based on regular expressions.
But it is easier to use and provides a metrics interface that is less fragile and fragmented, and more like an API.

Here, we will discuss Strimzi Metrics Reporter installation and configuration, and guide you on how to use it effectively to monitor your Kafka clusters.

### Key Features

> * Direct Prometheus Integration: Exposes Kafka metrics directly to Prometheus through an HTTP endpoint.

Author: Refactored

* Configurable Metrics Collection: Allows users to specify which metrics should be collected using a flexible allowlist.

#### Deploying Metrics Reporter
The first step is to install the Strimzi Metrics Reporter in your `Kafka` custom resource.
To do this, add the following:

```yaml
apiVersion: kafka.strimzi.io/v1beta1
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
  # ...
    metricsConfig:
      type: strimziMetricsReporter
  # ...  
```

By adding `type: strimziMetricsReporter` to the `metricsConfig` section of your `Kafka` custom resource, Strimzi will export a sensible set of default metrics.

> However, you can add you own custom values (providing the metrics are available) by filtering the metrics by name. 

Author: Well I just mean that the metrics have to be a valid Kafka metric but I suppose that's stating the obvious. Will remove

This is achieved by adding another field called `values`, and within that field, adding `allowList`.

> Then you can specify which metrics you want to collect as a regex list, where each entry is used to filter allowed metrics. 

Author: Refactored it to just: `Then you can specify which metrics you want to collect, where each entry is used to filter allowed metrics.`

> Then you can specify which metrics you want to collect as a regex list, where each entry is used to filter allowed metrics. 

Author: Yeah that makes it clearer. Added. Thanks @fvaleri

---
layout: post
title: "Using Strimzi Metrics Reporter to expose Kafka metrics in Prometheus format"
date: 2025-10-01
author: owen_corrigan
---

## Using Strimzi Metrics Reporter in your Kafka Cluster

> Apache Kafka® generates and exposes metrics that reflect how it is operating.

Author: Removed

The [Strimzi Metrics Reporter](https://github.com/strimzi/metrics-reporter) is a new project in Strimzi, still in early access, and it is available since Strimzi 0.48.0 release.
It implements the Kafka `MetricsReporter` interface to expose metrics via Prometheus.
Strimzi Metrics Reporter is designed from scratch as a Kafka plugin and allows you to configure it directly from the Kafka configuration.
That also means it is more efficient compared to other monitoring tools that run as sidecars or JVM agents.
Strimzi Metrics Reporter has fixed names for all metrics giving users much less flexibility than other tools which provide complex mapping rules based on regular expressions.
But it is easier to use and provides a metrics interface that is less fragile and fragmented, and more like an API.

Here, we will discuss Strimzi Metrics Reporter installation and configuration, and guide you on how to use it effectively to monitor your Kafka clusters.

### Key Features

> * "Native Prometheus support: The reporter exposes metrics in the Prometheus format through an HTTP endpoint, without relying on JMX.

Reviewer 2: No issues with _Native_ Prometheus support. But you have only the starting `"`.

> * Configurable Metrics Collection: Allows users to specify which metrics should be collected using a flexible allowlist.

Reviewer 2: Strictly speaking, you only expose the metrics. You do not collect them.

```suggestion
* Configurable Metrics Collection: Allows users to specify which metrics should be exposed using a flexible allowlist.
```

#### Deploying Metrics Reporter
The first step is to include the Strimzi Metrics Reporter in your `Kafka` custom resource.
To do this, add the following:

```yaml
apiVersion: kafka.strimzi.io/v1beta1
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
  # ...
    metricsConfig:
      type: strimziMetricsReporter
  # ...  
```

By adding `type: strimziMetricsReporter` to the `metricsConfig` section of your `Kafka` custom resource, Strimzi will export a sensible set of default metrics.
However, you can add you own custom values by filtering the metrics by name. 
This is achieved by adding another field called `values`, and within that field, adding `allowList`.
Then you can specify which metrics you want to collect, where each entry is used to filter allowed metrics.
You can put individual metrics names or use a regex to include a group of metrics and avoid having a long list.
For example:

```yaml
apiVersion: kafka.strimzi.io/v1beta1
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
  # ...
    metricsConfig:
      type: strimziMetricsReporter
      values:
        allowList:
          - "kafka_log.*"
          - "kafka_network.*"
  # ...  
```

> If you are adding Strimzi Metrics Reporter to an existing cluster, then you will see all your Kafka broker and Controller pods roll so that the Strimzi Metrics Reporter config is set in your `Kafa` resource.

Reviewer 3: ```suggestion
If you are adding Strimzi Metrics Reporter to an existing cluster, then you will see all your Kafka broker and Controller pods roll so that the Strimzi Metrics Reporter config is set in your `Kafka` resource.
```

> If you are adding Strimzi Metrics Reporter to an existing cluster, then you will see all your Kafka broker and Controller pods roll so that the Strimzi Metrics Reporter config can be added to your `Kafka` custom resource.

Author: @ppatierno I suppse I mean `If you are adding Strimzi Metrics Reporter to an existing cluster, then you will see all your Kafka broker and Controller pods roll when the Strimzi Metrics Reporter config is set in your `Kafa` resource.
`

Could it be worded like that??

If you would like to deploy your Kafka cluster with the Strimzi Metrics Reporter enabled from the start, you can use the examples in our [Strimzi Metrics Reporter Examples folder](https://github.com/strimzi/strimzi-kafka-operator/tree/0.48.0/examples/metrics/strimzi-metrics-reporter) by running the following command:

```bash
$ kubectl apply -f https://strimzi.io/examples/0.48.0/metrics/strimzi-metrics-reporter/kafka-metrics.yaml -n myproject
```

We also provide some [Grafana Dashboards](https://github.com/strimzi/strimzi-kafka-operator/tree/0.48.0/examples/metrics/strimzi-metrics-reporter/grafana-dashboards) to help you get started with visualizing your Kafka metrics in the Prometheus format.

> Strimzi Metrics Reporter can also be used with Kafka Connect and Kafka MirrorMaker 2, and we provide [examples](https://github.com/strimzi/strimzi-kafka-operator/tree/0.48.0/examples/metrics/strimzi-metrics-reporter) for these too.

Author: Added

---
layout: post

> title: "Using Strimzi Metrics Reporter to Expose Kafka Metrics in Prometheus Format"

Reviewer 4: do we need "Expose", "Metrics" and "Format" having capital first letter?

date: 2025-10-01

> author: Owen Corrigan

Reviewer 4: You need to add yourself in this list here https://github.com/strimzi/strimzi.github.io/blob/main/_data/authors.yml and then referring the entry from here.

Author: Done. Thanks for the info

---

> When deploying new Kafka cluster it may be one of your intentions to monitor the cluster using metrics.

Reviewer 5: Not sure about this sentence. It shouldn't be  "your intentions" to monitor a cluster, it's a must do. I wonder if we can drop that sentence as below you explain the important role of metrics.

> `Apache Kafka®` generates and exposes metrics that reflect how a system is operating. 

Reviewer 5: What about "how it is operating" instead of "how a system is operating", I mean it's exposing metrics about itself, not about other systems.

Reviewer 5: It's probably best to not format "Apache Kafka®" as code. Also use the full name on the first instance on the line above instead of here.

Author: Refactored

These metrics are useful for monitoring, troubleshooting, tuning, and capacity planning when it comes to running your Kafka cluster.
In essence, monitoring is crucial to ensure the health and performance of your Kafka clusters.

> `Prometheus` as a metrics monitoring solution has become increasingly common due to its popularity in the cloud-native ecosystem, and it is this format that we use to expose metrics.

Reviewer 5: I'd switch to active voice, ie
"`Prometheus` is a metrics monitoring solution that is increasingly common due to its popularity in the cloud-native ecosystem, and it is this format that we use to expose metrics."

Author: Done

> The [Strimzi Metrics Reporter](https://github.com/strimzi/metrics-reporter) project implements the Java `MetricsReporter` interface to expose metrics via `Prometheus` and it is available since Strimzi 0.48.0 release.

Reviewer 5: "the Java `MetricsReporter` interface" -> "the Kafka `MetricsReporter` interface"

Author: Silly mistake. Changed

> Up to this point, we have recommended using the `JMX Exporter` as a way to expose Kafka metrics, and we have provided some example yaml files in our [Strimzi Examples folder](https://github.com/strimzi/strimzi-kafka-operator/tree/main/packaging/examples/metrics).

Reviewer 5: Maybe link to the JMX Exporter project?

Reviewer 5: nit: yaml should be YAML

Author: Done

However, using the JMX Exporter adds unnecessary complexity.

> However, using the JMX Exporter introduces unnecessary complexity. 

Reviewer 5: Twice the same sentence?

> Metrics are first exposed by Kafka through JMX, and then a Java agent scrapes those JMX MBeans, applies mapping rules, and re-exposes them over HTTP for Prometheus. 

Reviewer 5: What about "then a Java agent scrapes" -> "then the JMX Exporter Java agent scrapes"

This extra indirection adds several drawbacks:

> * Overhead inside the JVM – The Java agent continuously scrapes and transforms metrics, consuming additional CPU and memory.

Reviewer 5: Not sure I would describe it as "continuously scrapes", I think this only happens when the HTTP metrics endpoint is queried. The main issue is that filtering happens when the endpoint is hit, while with the reporter the filtering happens when metrics are registered.

* Unnecessary double handling – Metrics are generated, registered as JMX MBeans, then scraped again, which adds latency and processing steps.
* Complex and fragile mappings – JMX Exporter’s regex-based rules can rename or restructure metrics. 

> While powerful, this makes it harder to treat metrics like an API, complicating debugging, dashboards, and integration with the Strimzi Operator.

Reviewer 5: +1, I actually think the JMX exporter is a pretty great project. It's able to handle any application using JMX. The Strimzi reporter is specifically designed for and usign Kafka specific APIs.

It's a different tradeoff. JMX Exporter is generic, Metrics reporter is specific. Being specific can be great for performance or specific features, but for example, it does not work with ZooKeeper or Cruise Control.

> While powerful, this makes it harder to treat metrics like an API, complicating debugging, dashboards, and integration with the Strimzi Operator.

Reviewer 4: +1

> While powerful, this makes it harder to treat metrics like an API, complicating debugging, dashboards, and integration with the Strimzi Operator.

Author: Removed this section

> The `Strimzi Metrics Reporter` exposes metrics in the Prometheus format directly over HTTP, significantly simplifying the setup and reducing resource consumption.

Reviewer 5: Again not sure we need the code formatting

> It is also a standalone MetricsReporter plugin that can be used with any Kafka Client deployment. 

Reviewer 5: Not only client, it works with non-Strimzi brokers and controllers too.

Author: Removed in the refactoring as part of Jakub's suggestion

Here, we will discuss Strimzi Metrics Reporter installation and configuration, and guide you on how to use it effectively to monitor your Kafka clusters.

> ## Key Features

Reviewer 2: You should use at least one-level lower headings ... so for example `###` here, `####` for the next one, etc. The website template does not translate the headers. So `h1` would be the website title, `h2` is the blog post title, and this would be something like `h3`. So add at least one `#` to all the headers an check the preview. It will look much better.

* Direct Prometheus Integration: Exposes Kafka metrics directly to Prometheus through an HTTP endpoint.
* Configurable Metrics Collection: Allows users to specify which metrics should be collected using a flexible allowlist.

### Deploying Metrics Reporter

> The first step is to install the metrics reporter in your `Kafka Custom Resource`. 

Reviewer 4: ```suggestion
The first step is to install the metrics reporter in your `Kafka` custom resource. 
```

```yaml
apiVersion: kafka.strimzi.io/v1beta1
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
  # ...
    metricsConfig:
      type: strimziMetricsReporter
  # ...  
```

By adding `type: strimziMetricsReporter` to the `metricsConfig` section of your Kafka CR, `Strimzi` will export a sensible set of default metrics.
However, you can add you own custom values (providing the metrics are available) by filtering the metrics by name. 
This is achieved by adding another field called `values` and within that field, adding `allowList`.
Then you can specify which metrics you want to collect as a regex list, where each entry is used to filter allowed metrics. 
For example:

```yaml
apiVersion: kafka.strimzi.io/v1beta1
kind: Kafka
metadata:
  name: my-cluster
spec:
  kafka:
  # ...
    metricsConfig:
      type: strimziMetricsReporter
      values:
        allowList:
          - "kafka_log.*"
          - "kafka_network.*"
  # ...  
```

If you are adding Metrics Reporter to an existing cluster (instead of starting a fresh cluster) then you will see all your Kafka broker and Controller pods roll so that the metrics reporter config can be added to your Kafka CR.

> If you would like to deploy your Kafka cluster with the Metrics Reporter enabled from the start, you can use the examples in our [Strimzi Metrics Reporter Examples folder](https://github.com/strimzi/strimzi-kafka-operator/tree/main/packaging/examples/metrics/strimzi-metrics-reporter) by running the following command from directory:

Reviewer 2: Here and in the next link ... you should definitely drop the packaging part from the path. You should also consider linking it to a particular tag maybe? At some point, we will likely switch to the MEtrics Reporter as the primary choice and the directory structure would change.

Author: Done

```bash

> $ kubectl apply -f kafka-metrics.yaml -n myproject

Reviewer 2: Would it make sense to use an URL? The `running the command from a directory` sounds weird.

```

> We have also provided some [Grafana Dashboards](https://github.com/strimzi/strimzi-kafka-operator/tree/main/packaging/examples/metrics/strimzi-metrics-reporter/grafana-dashboards) in the same directory to help you get started with visualizing your Kafka metrics in the Prometheus format.

Reviewer 5: Again active voice would read better -> "We also provide ..."

Author: Done

## Conclusion

> We have shown you how to deploy the `Strimzi Metrics Reporter` in your Kafka cluster and given some compelling reasons why you might want to use it in preference to the previously recommended `JMX Exporter`.

Reviewer 2: Here but also in other places - project names are not code. If you want to make them look differently, italics would probably look muhc better than fixed size font.

> As one recent reviewer wrote: `That strimzi metrics Reporter just shaved about 1.5 G of memory usage on our kafka pods. Good job`.

Reviewer 5: +1 for not including this

The `Strimzi Metrics Reporter` was recently [featured as part of StrimziCon 2025](https://www.youtube.com/watch?v=evKGEziQj54) which you may also find worth watching.

> We have also provided you with a practical demo video in which we go through the steps outlined above and hopefully this will encourage you to try out Strimzi Metrics Reporter for yourself.

Reviewer 4: I will provide you the link to the video when we are ready with this blog to be published.
